fl4health/feature_alignment/tab_features_info_encoder.py

```diff
@@ -0,0 +1,194 @@
+
+    Parameters
+    ----------
+    features_to_types: Dict[str, str]
+        Dictionary that maps each feature name to its type.
+        We consider four types in tabular data:
+        BINARY, ORDINAL, NUMERICAL, and STRING.
+    categories: Dict[str, List[Scalar]]
+        Dictionary that maps each ordinal feature to its categories.
+    target_info: TargetInfoEncoder
+        Information about the target column(s).
+    vocabulary: Dict[str, int]
+        Vocabulary of all the STRING columns.
+    default_fill_values: Dict[str, Scalar]
+        The default values used to fill in missing values.
+        Each of the four feature types has a default fill-in value, which is used to fill in all missing columns
+        of that type,
+        but each specific feature can also has its own default fill-in value.
+    """
+
+    def __init__(
+        self,
+        features_to_types: Dict[str, str],
+        categories: Dict[str, List[Scalar]],
+        target_info: TargetInfoEncoder,
+        vocabulary: Dict[str, int],
+        default_fill_values: Dict[str, Scalar] = DEFAULT_FILL_VALUES,
+    ) -> None:
+        self.features_to_types = features_to_types
+        self.categories = categories
+        self.target_info = target_info
+        self.vocabulary = vocabulary
+        self.default_fill_values = default_fill_values
+
+    def features_by_type(self, feature_type: str) -> List[str]:
+        return sorted([feature for feature, t in self.features_to_types.items() if t == feature_type])
+
+    def type_to_features(self) -> Dict[str, List[str]]:
+        return {feature_type: self.features_by_type(feature_type) for feature_type in FEATURE_TYPES}
+
+    def get_categories(self) -> Dict[str, List[Scalar]]:
+        return self.categories
+
+    def get_categories_list(self) -> List[List[Scalar]]:
+        return [self.get_categories()[feature_name] for feature_name in self.features_by_type(ORDINAL)]
+
+    def get_target(self) -> str:
+        return self.target_info.get_target()
+
+    def get_target_type(self) -> str:
+        return self.target_info.get_target_type()
+
+    def get_target_categories(self) -> List[Scalar]:
+        return self.target_info.get_target_categories()
+
+    def get_vocabulary(self) -> Dict[str, int]:
+        return self.vocabulary
+
+    def get_all_default_fill_values(self) -> Dict[str, Scalar]:
+        return self.default_fill_values
+
+    def get_default_fill_value(self, key: str) -> Scalar:
+        return self.default_fill_values[key]
+
+    def set_default_fill_value(self, key: str, value: Scalar) -> None:
+        self.default_fill_values[key] = value
+
+    @staticmethod
+    def encoder_from_dataframe(
+        df: pd.DataFrame,
+        id_column: str,
+        target_column: str,
+        default_fill_values: Dict[str, Scalar] = DEFAULT_FILL_VALUES,
+    ) -> "TabularFeaturesInfoEncoder":
+        features_list = sorted(df.columns.values.tolist())
+        features_list.remove(id_column)
```

Comment:
This requires that there be an `id_column`, but what if there isn't? Maybe we make it optional?

Reply:
A similar question for `TabularFeatures` also exists. I'm not super familiar with that class in CyclOps

Reply:
This is still a relevant comment after the new design. CyclOps requires there to be an id column, and I don't think it's unreasonable that we require there to be one as well, since we are working with pandas DataFrame here and the user could very easily add an id column if needed.

Reply:
I think that makes sense to me. I suppose the table index could always be used.